File: Matlab_code_with_GUI/Pso.m
classdef Pso < handle
   
    properties
        swarm;
        it_num;
        se_cy_num;
        me_cy_num;
        omega;
        c1;
        c2;
        c3;
        u;
        mud_slow;
        t;
    end
    methods
        function obj=Pso(seCyNum,meCyNum,omega,c1,c2,c3,domain,sub_swarm_size,part_num,u,t,c)
            obj.it_num=0;
            obj.se_cy_num=seCyNum;
            obj.me_cy_num=meCyNum;
            obj.omega=omega;
            obj.c1=c1;
            obj.c2=c2;
            obj.c3=c3;
            obj.u=u;
            obj.t=t;
            obj.mud_slow=(1.0/1.2);
            obj.swarm=Swarm(domain,sub_swarm_size,part_num,u,t,c);
            obj.swarm.fit();
        end
        
        function v=get_best_fit(PSO)
           v=PSO.swarm.best_particle.best_fit;
        end
        
        function arg=get_best_fit_arg(PSO)
           arg=PSO.swarm.best_particle.best_arg;
        end

        function it=get_it_num(PSO)
           it=PSO.it_num; 
        end
        
        function PSO=do_next_it(PSO)
           for i=1:PSO.me_cy_num
              PSO.do_me(); 
           end
           for i=1:PSO.se_cy_num
               PSO.do_se();
           end
           PSO.it_num=PSO.it_num+1;
        end
        
        function do_se(PSO)
           for i=1:PSO.swarm.sub_swarm_size
               ss=PSO.swarm.sub_swarm(i);
               for j=1:PSO.swarm.sub_swarm(i).size
                   p=ss.particles(j);
                   while true
                       r1=rand;
                       r2=rand;
                       r3=rand;
                       v=PSO.omega*p.vel+PSO.c1*r1*(p.best_arg-p.arg)+PSO.c2*r2*(ss.best_particle.best_arg-p.arg)+PSO.c3*r3*(PSO.swarm.best_particle.best_arg-p.arg);
                       if(PSO.swarm.is_in_domain(p.arg+v))
                           p.fit(v);
                           break;
                       else
                           p.vel=p.vel*PSO.mud_slow;
                       end
                   end
               end
           end
           PSO.swarm.fit();
        end
       
       function do_me(PSO)
           for i=1:PSO.swarm.sub_swarm_size
               ss=PSO.swarm.sub_swarm(i);
               for j=1:ss.size
                   p=ss.particles(j);
                   while true
                       r1=rand;
                       r2=rand;
                       v=PSO.omega*p.vel+PSO.c1*r1*(p.best_arg-p.arg)+PSO.c2*r2*(ss.best_particle.best_arg-p.arg);
                       if(PSO.swarm.is_in_domain(p.arg+v))
                           p.fit(v);
                           break;
                       else
                           p.vel=p.vel*PSO.mud_slow;
                       end
                   end
               end
           end
           PSO.swarm.fit();
       end
    end
    
end

File: c++_code(dirty)/Pso.cpp
#include "Pso.h"

Pso::Pso(int seCyNum,int meCyNum, double omega,double c1,double c2,double c3,std::vector<Vel_pos>domain,int sub_swarm_size, int part_num,std::vector<Vel_pos> u, int t, int c, bool f):
    _se_cy_num(seCyNum), _me_cy_num(meCyNum),_omega(omega),_c1(c1),_c2(c2),_c3(c3),mud_slow(1.0/1.2)
{//std::cout<<"zaczynam konstruktor PSO"<<std::endl;
    _it_num=0;
    _u=u;
    _f=f;
    _swarm=Swarm(domain,sub_swarm_size,part_num,u,t,c,f);
    _swarm.fit();
//std::cout<<"koncze konstruktor PSO"<<std::endl;
}

Pso::~Pso()
{
}
double Pso::get_best_fit(){
    return _swarm.get_best_particle().get_best_fit();
}
Vel_pos Pso::get_best_fit_arg(){
    return _swarm.get_best_particle_as_cp().get_best_fit_arg();
}

int Pso::get_it_num(){
    return _it_num;
}

void Pso::do_next_it(){
    for(int i=0; i<_me_cy_num; i++)
        _do_me();
    for(int i=0; i<_se_cy_num; i++)
        _do_se();
    _it_num++;
    //std::cout<<"najlepiej dopasowana:  fit: "<<_swarm.get_best_particle().get_best_fit()<<std::endl;
}

void Pso::_do_se(){
    Sub_swarm* sub_swarm;
    Particle* particle;
    for(int i=0; i<_swarm.get_sub_swarm_size();i++){
        sub_swarm=_swarm.get_sub_swarm(i);//ZWRACAC REFERENCJE
        for(int j=0; j<sub_swarm->get_size(); j++){
            particle=sub_swarm->get_particle(j);//ZWRACAC REFERENCJE
            while(true){
                double r1=(double)rand()/(double)RAND_MAX;
                double r2=(double)rand()/(double)RAND_MAX;
                double r3=(double)rand()/(double)RAND_MAX;
                Vel_pos v=_omega*particle->get_vel()+(_c1*r1*(particle->get_best_fit_arg()-particle->get_arg()))+(_c2*r2*(sub_swarm->get_best_particle().get_best_fit_arg()-particle->get_arg()))+(_c3*r3*(_swarm.get_best_particle().get_best_fit_arg()-particle->get_arg()));
                if(_swarm.is_in_domain(particle->get_arg()+v)){//std::cout<<"weszlem a nie widzisz"<<std::endl;
                    particle->fit(v);//zmienic wszystko, fit moze zajac sie zmiana parametrow
                    break;
                }
                else{
                    particle->set_vel(mud_slow*particle->get_vel());
                }
            }
        }
    }
    _swarm.fit();
}
void Pso::_do_me(){
    Sub_swarm* sub_swarm;
    Particle* particle;
    for(int i=0; i<_swarm.get_sub_swarm_size();i++){
        sub_swarm=_swarm.get_sub_swarm(i);//ZWRACAC REFERENCJE
        for(int j=0; j<sub_swarm->get_size(); j++){
            particle=sub_swarm->get_particle(j);//ZWRACAC REFERENCJE
            while(true){
                double r1=(double)rand()/(double)RAND_MAX;
                double r2=(double)rand()/(double)RAND_MAX;
                Vel_pos v=_omega*particle->get_vel()+(_c1*r1*(particle->get_best_fit_arg()-particle->get_arg()))+(_c2*r2*(sub_swarm->get_best_particle().get_best_fit_arg()-particle->get_arg()));//jed
                if(_swarm.is_in_domain(particle->get_arg()+v)){//std::cout<<"dzialam"<<std::endl;//std::cout<<"weszlem a nie widzisz"<<std::endl;
                    particle->fit(v);//zmienic wszystko, fit moze zajac sie zmiana parametrow
                    break;
                }
                else{
                    particle->set_vel(mud_slow*particle->get_vel());
                }
            }
        }
    }
    _swarm.fit();
}
